charpick/charpick.c

```diff
@@ -350,478 +362,540 @@ chooser_button_clicked (GtkButton *button, charpick_data *curr_data)
-	gtk_box_set_homogeneous (GTK_BOX (row_box[i]), TRUE);
-  	gtk_box_pack_start (GTK_BOX (button_box), row_box[i], TRUE, TRUE, 0);
-  }
-  
-  for (i = 0; i <len; i++) {  	
-  	int delta = len/size_ratio;
-  	int index;
-  
-	if (delta > 0)
-	  	index = i / delta;
-	else
-		index = i;
+        toggle_button[i] = gtk_toggle_button_new_with_label (label);
+        atk_desc =  g_strdup_printf (_("insert special character %s"),
+                                     label);
+
+        set_atk_name_description (toggle_button[i],
+                                  NULL,
+                                  atk_desc);
+        g_free (atk_desc);
+        gtk_widget_show (toggle_button[i]);
+        gtk_button_set_relief (GTK_BUTTON (toggle_button[i]),
+                               GTK_RELIEF_NONE);
+
+        /* FIXME : evil hack (see force_no_button_padding) */
+        force_no_button_padding (toggle_button[i]);
+
+        gtk_widget_set_tooltip_text (toggle_button[i],
+                                     name);
+        g_free (name);
+
+        gtk_widget_get_preferred_size (toggle_button[i],
+                                       NULL,
+                                       &req);
+
+        max_width = MAX (max_width, req.width);
+        max_height = MAX (max_height, req.height-2);
+
+        g_object_set_data (G_OBJECT (toggle_button[i]),
+                           "unichar", 
+                           GINT_TO_POINTER (g_utf8_get_char (label)));
+
+        g_signal_connect (G_OBJECT (toggle_button[i]),
+                          "toggled",
+                          G_CALLBACK (toggle_button_toggled_cb),
+                          p_curr_data);
+
+        g_signal_connect (G_OBJECT (toggle_button[i]),
+                          "button_press_event", 
+                          G_CALLBACK (button_press_hack),
+                          p_curr_data->applet);
+    }
 
-	index = CLAMP (index, 0, size_ratio-1);	
-  	gtk_box_pack_start (GTK_BOX (row_box[index]), toggle_button[i], TRUE, TRUE, 0);
-  }
+    if (p_curr_data->panel_vertical) {
+        size_ratio = p_curr_data->panel_size / max_width;
+        button_box = gtk_box_new (GTK_ORIENTATION_HORIZONTAL,
+                                  0);
+    } else {
+        size_ratio = p_curr_data->panel_size / max_height;
+        button_box = gtk_box_new (GTK_ORIENTATION_VERTICAL,
+                                  0);
+    }
 
-  g_free (toggle_button);
-  g_free (row_box);
+    gtk_box_set_homogeneous (GTK_BOX (button_box), TRUE);
+    gtk_box_pack_start (GTK_BOX (box),
+                        button_box,
+                        TRUE, TRUE, 0);
+
+    size_ratio = MAX (size_ratio, 1);
+    row_box = g_new0 (GtkWidget *, size_ratio);
+    for (i=0; i < size_ratio; i++) {
+        if (!p_curr_data->panel_vertical) row_box[i] = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 0);
+        else row_box[i] = gtk_box_new (GTK_ORIENTATION_VERTICAL, 0);
```

Comment:
the previous two lines cannot be read easily, you may need to use a ternary operator or to rewrite the if-else block using new lines for bodies and swapping the if/else blocks

Reply:
Should be fine now.